File: 208.implement-trie-prefix-tree.py
class Trie:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.dict = dict()

    def insert(self, word: str) -> None:
        """
        Inserts a word into the trie.
        """
        r = self.dict
        for i in word:
            if i not in r.keys():
                r[i] = dict()
            r = r[i]
        r['$'] = None

    def search(self, word: str) -> bool:
        """
        Returns if the word is in the trie.
        """
        r = self.dict
        for i in word:
            if i not in r:
                return False
            r = r[i]
        return '$' in r

    def startsWith(self, prefix: str) -> bool:
        """
        Returns if there is any word in the trie that starts with the given prefix.
        """
        r = self.dict
        for i in prefix:
            if i not in r:
                return False
            r = r[i]
        return True


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

File: 208.implement-trie-prefix-tree.java
class Trie {

    class TrieNode {
        public Boolean isWord;
        public HashMap<Character, TrieNode> m;

        public TrieNode() {
            this.isWord = false;
            this.m = new HashMap<>();
        }
    }

    public TrieNode head;

    public Trie() {
        this.head = new TrieNode();
    }

    public void insert(String word) {
        TrieNode head = this.head;
        for (int i = 0; i < word.length(); i++) {
            char c = word.charAt(i);
            if (!head.m.containsKey(c)) {
                head.m.put(c, new TrieNode());
            }
            head = head.m.get(c);
        }
        head.isWord = true;
    }

    public boolean search(String word) {
        TrieNode head = this.head;
        for (int i = 0; i < word.length(); i++) {
            char c = word.charAt(i);
            if (!head.m.containsKey(c)) {
                return false;
            }
            head = head.m.get(c);
        }
        return head.isWord;
    }

    public boolean startsWith(String prefix) {
        TrieNode head = this.head;
        for (int i = 0; i < prefix.length(); i++) {
            char c = prefix.charAt(i);
            if (!head.m.containsKey(c)) {
                return false;
            }
            head = head.m.get(c);
        }
        return this.isPrefix(head);
    }

    private boolean isPrefix(TrieNode head) {
        if (head.isWord) {
            return true;
        }
        for (Map.Entry<Character, TrieNode> entry : head.m.entrySet()) {
            if (this.isPrefix(entry.getValue())) {
                return true;
            }
        }
        return false;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */